File: 063. Unique Paths II—Dynamic Programming/Unique Paths II.py
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        if not len(obstacleGrid) or not len(obstacleGrid[0]): return 0
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        dp = [[0] * (n+1) for _ in xrange(m+1)]
        dp[0][1] = 1
        for r in xrange(1, m+1):
            for c in xrange(1, n+1):
                if obstacleGrid[r-1][c-1] == 0:
                    dp[r][c] = dp[r-1][c] + dp[r][c-1]
        return dp[-1][-1]

File: 063. Unique Paths II—Dynamic Programming/Unique Paths II.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size(); //行数
        int n = obstacleGrid[0].size();//列数
        int p[m][n];
       
       //第一列赋值 
        int k = 0;
        while(k < m && obstacleGrid[k][0] != 1)
            p[k++][0] = 1;
       //如果遇到了障碍物则它及其后面的值都为0
        while(k < m)
            p[k++][0] = 0;
        
        //第一行赋值
        k = 0;
        while(k < n && obstacleGrid[0][k] != 1)
            p[0][k++] = 1;
        while(k < n)
            p[0][k++] = 0;
        
        for(int i = 1; i < m; i++)
            for(int j = 1; j < n; j++){
                if(obstacleGrid[i][j] == 1) //如果遇到障碍物，则该位置的值为0
                    p[i][j] = 0;
                else
                    p[i][j] = p[i - 1][j] + p[i][j - 1];
            }
        return p[m-1][n-1];
    }
};
